index.js

```diff
@@ -101,50 +102,139 @@ function mapChars(domainName, useSTD3, processingOption) {
-
-
-function validateLabel(label, processingOption) {
+function validateLabel(label, processingOption, checkHyphens, checkBidi, checkJoiners) {
   if (label.substring(0, 4) === "xn--") {
-    label = punycode.decode(label.substring(4));
+    try {
+      label = punycode.decode(label.substring(4));
+    } catch (err) {
+      return {
+        label,
+        error: true
+      };
+    }
     processingOption = PROCESSING_OPTIONS.NONTRANSITIONAL;
   }
 
-  let error = false;
+  if (normalize(label) !== label) {
+    return {
+      label,
+      error: true
+    };
+  }
+
+  if (checkHyphens) {
+    if (label[2] === "-" && label[3] === "-" ||
+        label[0] === "-" || label[label.length - 1] === "-") {
+      return {
+        label,
+        error: true
+      };
+    }
+  }
 
-  if (normalize(label) !== label ||
-      (label[3] === "-" && label[4] === "-") ||
-      label[0] === "-" || label[label.length - 1] === "-" ||
-      label.includes(".") ||
-      label.search(combiningMarksRegex) === 0) {
-    error = true;
+  if (label.includes(".") ||
+      label.search(regexes.combiningMarks) === 0) {
+    return {
+      label,
+      error: true
+    };
   }
 
   const len = countSymbols(label);
   for (let i = 0; i < len; ++i) {
-    const status = findStatus(label.codePointAt(i));
+    const cp = label.codePointAt(i);
+    const status = findStatus(cp);
     if ((processingOption === PROCESSING_OPTIONS.TRANSITIONAL && status[1] !== "valid") ||
         (processingOption === PROCESSING_OPTIONS.NONTRANSITIONAL &&
          status[1] !== "valid" && status[1] !== "deviation")) {
-      error = true;
-      break;
+      return {
+        label,
+        error: true
+      };
+    }
+    if (cp > 0xFFFF) {
+      i++;
+    }
+  }
+
+  if (checkJoiners) {
+    let last = 0;
+    for (let i = 0; i < len; ++i) {
+      if (label[i] === "\u200C" || label[i] === "\u200D") {
+        if (i > 0) {
+          if (regexes.combiningClassVirama.test(label[i - 1])) {
+            continue;
+          }
+          if (label[i] === "\u200C") {
+            // TODO: make this more efficient
+            const next = label.indexOf("\u200C", i + 1);
+            const test = next < 0 ? label.slice(last) : label.slice(last, next);
+            if (regexes.validZWNJ.test(test)) {
+              last = i + 1;
+              continue;
+            }
+          }
+        }
+        return {
+          label,
+          error: true
+        };
+      }
+    }
+  }
+
+  if (checkBidi) {
+    let rtl;
+
+    // 1
+    const first = label.codePointAt(0);
+    if (regexes.bidiS1LTR.check(first)) {
+      rtl = false;
+    } else if (regexes.bidiS1RTL.check(first)) {
+      rtl = true;
+    } else {
+      return {
+        label,
+        error: true
+      };
+    }
+
+    if (rtl) {
```

Comment:
I followed this order since it's also the order used by the spec, but I'm okay with that.